### ISSUE TITLE
Unable to use AWS SSO config with packer

### ISSUE THREAD

**Issue by baronnenate** [hashicorp/packer#10525]
When filing a bug, please include the following headings if possible. Any
example text in this template can be deleted.

#### Overview of the Issue

From the AWS CLI I can configure a profile using `aws configure sso --profile sandbox` - this is handy as I can run in the context of my own account and test packer in the sandbox environment - however - packer doesn't appear to have a way to honour that sso profile configuration as it requires access keys.

#### Reproduction Steps

aws configure sso --profile sandbox
packer build -var-file .\variables.json .\sandbox.json

### Packer version

1.6.6

### Simplified Packer Buildfile

If the file is longer than a few dozen lines, please include the URL to the
{
    "builders": [
        {
            "type": "amazon-ebs",
            "profile": "{{user `aws_profile`}}",
            "region": "{{user `region`}}",
            "iam_instance_profile": "{{user `iam_instance_profile`}}",

### Operating system and Environment details

Windows 10 - PowerShell 7 - AWS CLI - Visual Studio Code

### Log Fragments and crash.log files

amazon-ebs: output will be in this color.

Build 'amazon-ebs' errored after 304 milliseconds 547 microseconds: no valid credential sources for  found.

Please see
for more information about providing credentials.

Error: NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors

==> Wait completed after 304 milliseconds 547 microseconds

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: no valid credential sources for  found.

Please see
for more information about providing credentials.

**Comment by SwampDragons** [comment]
My first pieces of advice is to make sure that no other credential tokens are set inside the Packer template -- i.e. no "access_key" or "secret_key" or "session_token" values in the build object.

Your goal is to use this method: https://www.packer.io/docs/builders/amazon#shared-credentials-file so if your shared credentials profile is not in the default location, make sure the env var documented there is set so the Amazon SDK knows where to look.

Can you verify that you're doing both these things?  I'd be really surprised if this is not working since we use the SDK provided by AWS to do authentication.

**Comment by baronnenate** [comment]
> My first pieces of advice is to make sure that no other credential tokens are set inside the Packer template -- i.e. no "access_key" or "secret_key" or "session_token" values in the build object.
> 
> Your goal is to use this method: https://www.packer.io/docs/builders/amazon#shared-credentials-file so if your shared credentials profile is not in the default location, make sure the env var documented there is set so the Amazon SDK knows where to look.
> 
> Can you verify that you're doing both these things? I'd be really surprised if this is not working since we use the SDK provided by AWS to do authentication.

Thanks for the reply.  I can confirm there is no reference to access_key or secret_key in the build json file - the only thing I have referenced is the profile. The problem with the reference to the shared credentials file is that with the AWS SSO, there is no credentials file stored in `%USERPROFILE%.aws\credential` - there is a config file which looks like this:
```
[profile sandbox]
sso_start_url = https://**redacted**.awsapps.com/start#
sso_region = eu-west-2
sso_account_id = **redacted**
sso_role_name = **redacted**
region = eu-west-2
output = json
```

**Comment by SwampDragons** [comment]
ah, looks like this _is_ an issue with the sdk: https://github.com/aws/aws-sdk-go/issues/3186

From that poster's comment: 

> If you use AWS CLI 2.0 SSO integration, you don't end up with credentials stored either in environment variables or a shared credentials file in $HOME/.aws/credentials. What you do get is a json file in $HOME/.aws/sso/cache/ with an access token. The first time you run AWS CLI command it produces another json file in $HOME/.aws/cli/cache that contains an AccessKeyId, SecretAccessKey and SessionToken that (if converted into environment variables or a credentials file) can be used to run CLI commands. These are needed by aws-iam-authenticator to connect to k8s.

If this is true, then maybe as a workaround while we wait for the SDK to actually support its own SSO, you can run a get-images query or something to "prime" the cli and make it produce a credentials file that Packer can read?

**Comment by SwampDragons** [comment]
another lead from [these sso docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sso.html): 
> At this point, you have a profile that you can use to request temporary credentials. You must use the aws sso login command to actually request and retrieve the temporary credentials needed to run commands. 

Try calling `aws sso login` before running Packer

Looks like this page has detailed instructions:https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sso.html#sso-using-profile

So even though Packer can't use your sso credentials to log in and retrieve credentials, it looks like it should be able to use credentials you retrieve for it by logging in.

**Comment by baronnenate** [comment]
thanks I've tried the `aws sso login` before running the packer build command but unfortunately no dice. It looks to me like the workarounds are a bit fiddly - various people have created their own python packages that you can install that wrap around this issue. I'd prefer if packer was able to natively work with the SSO as expected. 
https://github.com/benkehoe/aws-sso-credential-process
https://github.com/benkehoe/aws-sso-util
https://github.com/victorskl/yawsso

**Comment by baronnenate** [comment]
note I also tried 'priming' the CLI as you suggested by issuing a command eg. `aws ec2 describe-images` --profile sandbox

**Comment by SwampDragons** [comment]
I'd like it to work too, but I'm not sure the Packer team is going to have a chance to look into this before the Amazon team fixes it in their golang SDK. If it's a major blocker for you, you're welcome to take a look at it yourself -- the code for authenticating to Amazon is [here](https://github.com/hashicorp/packer/blob/master/builder/amazon/common/access_config.go); the challenge there is going to be making sure that what works for you works for other users as well, and making sure that SSO logins don't cause the Packer build to hang indefinitely waiting for authentication. Otherwise, you're going to have to figure out where your sso login call creates the temporary credentials and make sure Packer knows where to look for them.

**Comment by baronnenate** [comment]
thanks. it's not a blocker - it's just a "would be nice". Sadly I'm not a developer so wouldn't have a clue when it comes to looking at the code, let alone contributing!

**Comment by lpkirby** [comment]
As per this document: https://aws.amazon.com/blogs/developer/aws-sso-support-in-the-aws-sdk-for-go/
the golang SDK seems to be updated to now support AWS SSO as of the first part of February.  I just updated to Packer 1.7.0, released 2 days ago, and AWS SSO still doesn't work.

I'm not that familiar with golang.  Is there some version dependency or something that needs to be updated to bring in the latest golang aws sdk?

Incidentally, AWS SSO works nicely in Terraform in case that's helpful.

**Comment by SwampDragons** [comment]
Yeah we probably need to bump the SDK dependency, but it may be a necessary-but-not-sufficient type thing. I don't use AWS sso -- if I bump the SDK and rebuild, can someone affected test the Packer build to see whether that's all we need to do?

**Comment by lpkirby** [comment]
@SwampDragons I'd be happy to test.

**Comment by SwampDragons** [comment]
Thanks! You can pull a test build using the latest SDK off our CI server here: https://app.circleci.com/pipelines/github/hashicorp/packer/9298/workflows/e9021d7e-8628-4617-80c9-cee66f46a0f4/jobs/110905/artifacts

**Comment by lpkirby** [comment]
@SwampDragons Thank you for jumping on that so quickly!  :)  I can confirm that updated fixed AWS SSO support.  It works now.

In case it matters, I wasn't able to download the artifact directly from circleci.  It wanted a lot more Github permissions than I was willing to grant it (especially to other orgs) so I built the binary locally from your update_amazon_sdk branch.

Thanks again.

**Comment by SwampDragons** [comment]
Perfect!  

We put the artifacts on circle for uses who don't feel comfortable building golang themselves; I totally support you building locally to protect your information. I'll merge that PR and SSO support will be officially available in the 1.7.1 release
